kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/blockchain/BlockchainSourceImpl.java

```diff
@@ -0,0 +1,277 @@
+    private final MainThreadExecutor mainThread = new MainThreadExecutor();
+
+    private String appID;
+    private static final int MEMO_FORMAT_VERSION = 1;
+    private static final String MEMO_DELIMITER = "-";
+    private static final String MEMO_FORMAT =
+        "%d" + MEMO_DELIMITER + "%s" + MEMO_DELIMITER + "%s"; // version-appID-orderID
+
+    private static final int APP_ID_INDEX = 1;
+    private static final int ORDER_ID_INDEX = 2;
+    private static final int MEMO_SPLIT_LENGTH = 3;
+
+    private BlockchainSourceImpl(@NonNull final Context context, @NonNull BlockchainSource.Local local)
+        throws InitializeException {
+        this.local = local;
+        this.kinClient = new KinClient(context, Network.NETWORK_PRIVATE_TEST.getProvider());
+        createKinAccountIfNeeded();
+        getCurrentBalance();
+    }
+
+    public static void init(@NonNull final Context context, @NonNull BlockchainSource.Local local) throws InitializeException {
+        if (instance == null) {
+            synchronized (BlockchainSourceImpl.class) {
+                if (instance == null) {
+                    instance = new BlockchainSourceImpl(context, local);
+                }
+            }
+        }
+    }
+
+    public static BlockchainSourceImpl getInstance() {
+        return instance;
+    }
+
+    private void createKinAccountIfNeeded() throws InitializeException {
+        account = kinClient.getAccount(0);
+        if (account == null) {
+            try {
+                account = kinClient.addAccount();
+                startAccountCreationListener();
+            } catch (CreateAccountException e) {
+                throw new InitializeException(e.getMessage());
+            }
+        }
+    }
+
+    @Override
+    public void setAppID(String appID) {
+        Log.d(TAG, "setAppID: " + appID);
+        if (!TextUtils.isEmpty(appID)) {
+            this.appID = appID;
+        }
+    }
+
+    @Override
+    public void sendTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount,
+        @NonNull final String orderID) {
+        account.sendTransaction(publicAddress, amount, generateMemo(orderID)).run(
+            new ResultCallback<TransactionId>() {
+                @Override
+                public void onResult(TransactionId result) {
+                    Log.d(TAG, "sendTransaction onResult: " + result.id());
+                }
+
+                @Override
+                public void onError(Exception e) {
+                    completedPayment.setValue(new Payment(orderID, false, e.getMessage()));
+                    Log.d(TAG, "sendTransaction onError: " + e.getMessage());
+                }
+            });
+    }
+
+    @SuppressLint("DefaultLocale")
+    private String generateMemo(@NonNull final String orderID) {
+        return String.format(MEMO_FORMAT, MEMO_FORMAT_VERSION, appID, orderID);
+    }
+
+
+    private void getCurrentBalance() {
+        getBalance();
```

Comment:
why do we need to call getBalance()?

Reply:
getBalance with no Callback argument is taken from the local data.
getBalance with a Callback gets it from the blockchain network.

So first, we want the locally cached balance and then we refresh it, all of this happens internally it's not something we expose.